### PR TITLE
ENT-5865: feat: Added category and sub-category for course skills.

### PR DESCRIPTION
**Pull request by saleem-latif**
**Jira Ticket:** [ENT-5868](https://2u-internal.atlassian.net/browse/ENT-5865)

__Description:__
This PR adds new models and fields for handling category and subcategory data for skills. A management command is included that will execute weekly to fetch and populate skill category and subcatefory.

**Merge checklist:**
- [x] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `make upgrade && make requirements` have been run to regenerate requirements
- [x] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
- [x] [Version](https://github.com/edx/taxonomy-connector/blob/master/taxonomy/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/taxonomy-connector/blob/master/CHANGELOG.rst) record added

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/taxonomy-connector/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/taxonomy-connector/actions), verify version has been pushed to [PyPI](https://pypi.org/project/taxonomy-connector/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [course-discovery](https://github.com/edx/course-discovery) to upgrade dependencies (including taxonomy-connector)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in course-discovery will look for the latest version in PyPi.

### CODE REVIEW

**Review by muneebGH — APPROVED** [review]
LGTM 🥳

File: taxonomy/emsi/client.py [hunk 1]
```diff
@@ -115,14 +116,41 @@ class EMSISkillsApiClient(JwtEMSIApiClient):
     Object builds an API client to make calls to get the skills from course text data.
     """
 
-    API_BASE_URL = JwtEMSIApiClient.API_BASE_URL + '/skills/versions/7.35'
+    API_BASE_URL = JwtEMSIApiClient.API_BASE_URL + '/skills/versions/8.9'
```
Comment (HammadAhmadWaqas) [inline taxonomy/emsi/client.py:119]: can we use urljoin here?

File: taxonomy/emsi/client.py [hunk 2]
```diff
@@ -115,14 +116,41 @@ class EMSISkillsApiClient(JwtEMSIApiClient):
...
+
+        We will be using this method to populate skill category and subcategory.
+
+        Arguments:
+             skill_id (str): Skill external id, this is the id that comes from EMSI. example: 'KS124P772D5HNCJGGQ05'
+
+        Returns:
+            (dict): A dictionary containing the skill details.
+        """
+        url = 'skills/{skill_id}'.format(
```
Comment (HammadAhmadWaqas) [inline taxonomy/emsi/client.py:140]: here as well

File: taxonomy/management/commands/fetch_skill_details.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            )
+            skill.category = skill_category
+            skill_subcategory, _ = SkillSubCategory.objects.get_or_create(
+                id=subcategory['id'],
+                defaults={'name': subcategory['name']},
+                category=skill_category,
+            )
+            skill.subcategory = skill_subcategory
+
+        skill.save()
```
Comment (HammadAhmadWaqas) [inline taxonomy/management/commands/fetch_skill_details.py:53]: is not this skill.save() should move under the if statement?

File: taxonomy/management/commands/fetch_skill_details.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        Persist the skill category and subcategory data in the database.
+
+        Arguments:
+            skill (Skill): Skill instance whose category and subcategory needs to be added/updated.
+            skill_data (dict): A dictionary containing skill category and subcategory.
+        """
+        LOGGER.info('Updating category and subcategory data for skill external id {}.'.format(skill.external_id))
+        category = skill_data.get('category')
+        subcategory = skill_data.get('subcategory')
+        if category and subcategory:
```
Comment (HammadAhmadWaqas) [inline taxonomy/management/commands/fetch_skill_details.py:40]: if catergory is there and only subcatergory is missing, do We want to save the category to the skill?

**Review by HammadAhmadWaqas — COMMENTED** [review]
posted some questions

File: taxonomy/management/commands/fetch_skill_details.py (hunk 4 above)
Comment (muhammad-ammar) [reply taxonomy/management/commands/fetch_skill_details.py:40]: it may be possible that there is no subcategory for s skill?

File: taxonomy/emsi/parsers/skill_parsers.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+"""
+Module that contains utility methods and classes for parsing EMSI responses.
+"""
+INVALID_NAMES = {'NULL', 'NONE', ''}
+
+
+class SkillDataParser:
+    """
+    Parser for processing/parsing EMSI responses returned by th skills API (https://api.emsidata.com/apis/skills).
```
Comment (sameenfatima78) [inline taxonomy/emsi/parsers/skill_parsers.py:9]: nit: th -> the

File: taxonomy/emsi/parsers/skill_parsers.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            (dict): A dictionary containing at-least following 2 keys
+                1. category  -> value against this key will be a dict containing `id` and `name`.
+                2. subcategory -> value against this key will be a dict containing `id` and `name`.
+        """
+        category = self.data.get('category')
+        subcategory = self.data.get('subcategory')
+
+        if category is None or category.get('name') is None or category.get('name').upper() in INVALID_NAMES:
+            category = None
+            subcategory = None  # We can not hav a non-null subcategory with null category.
```
Comment (sameenfatima78) [inline taxonomy/emsi/parsers/skill_parsers.py:35]: nit: hav -> have

File: taxonomy/emsi/client.py [hunk 7]
```diff
@@ -148,7 +176,14 @@ def get_course_skills(self, course_text_data):
             raise TaxonomyAPIError('Error while fetching course skills.') from error
 
     @staticmethod
-    def traverse_data(response):
+    def traverse_skill_detail_data(response):
+        """
+        Transform data to a more useful format.
+        """
+        return response
+
```
Comment (sameenfatima78) [inline taxonomy/emsi/client.py:184]: Can we discard this method and return the response directly?

File: taxonomy/management/commands/fetch_skill_details.py (hunk 4 above)
Comment (sameenfatima78) [reply taxonomy/management/commands/fetch_skill_details.py:40]: Do we not want to save the category in case sub-category doesn't exist? If yes, why not?

File: tests/emsi/parsers/test_skill_parsers.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import ddt
+
+from taxonomy.emsi.parsers.skill_parsers import SkillDataParser
+from test_utils.testcase import TaxonomyTestCase
+
+
+@ddt.ddt
+class TestSkillDataParser(TaxonomyTestCase):
+    """
+    Validate that JWT token are fetched and cached appropriately.
```
Comment (sameenfatima78) [inline tests/emsi/parsers/test_skill_parsers.py:14]: nit: this docstring doesn't seem relevant to the testcase class

**Review by sameenfatima78 — COMMENTED** [review]
A few questions and suggestions

File: taxonomy/emsi/client.py (hunk 2 above)
Comment (saleem-latif) [reply taxonomy/emsi/client.py:140]: I do not think it is a good idea to use `urljoin` here since this is actually url path. I have renamed the variable to `path`.

File: taxonomy/management/commands/fetch_skill_details.py (hunk 4 above)
Comment (saleem-latif) [reply taxonomy/management/commands/fetch_skill_details.py:40]: I have updated the logic to save category regardless of subcategory.

File: tests/emsi/parsers/test_skill_parsers.py (hunk 8 above)
Comment (saleem-latif) [reply tests/emsi/parsers/test_skill_parsers.py:14]: Copy-Paste issue :)

File: taxonomy/management/commands/fetch_skill_details.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    def _fetch_skill_category_and_sub_category(self):
+        """
+        Fetch skill category and subcategory data from EMSI and update the database accordingly.
+        """
+        client = EMSISkillsApiClient()
+        try:
+            skills = Skill.objects.filter(Q(category__isnull=True) | Q(subcategory__isnull=True))
+            for chunked_skills in chunked_queryset(skills, chunk_size=100):
+                for skill in chunked_skills:
+                    response = client.get_skill_details(skill_id=skill.external_id)
```
Comment (muhammad-ammar) [inline taxonomy/management/commands/fetch_skill_details.py:66]: Is it possible to get details of multiple skills at once instead of making a call for every skill?

**Review by sameenfatima78 — APPROVED** [review]

File: taxonomy/management/commands/fetch_skill_details.py (hunk 9 above)
Comment (saleem-latif) [reply taxonomy/management/commands/fetch_skill_details.py:66]: I tried to find that endpoint but could not find it in EMSI API documentation.
